libass/ass_font.c

```diff
@@ -35,6 +42,197 @@
 #include "ass_utils.h"
 #include "ass_shaper.h"
 
+
+/**
+ * Convert a UCS-4 code unit to a packed uint32_t in given multibyte encoding
+ */
+#if defined(_WIN32)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    uint32_t result = codepoint;
+    uint16_t input_buffer[2];
+    char output_buffer[4];
+    size_t inbuf_size;
+    UINT codepage;
+    BOOL conversion_fail;
+
+    // encode surrogate pair, assuming codepoint > 0 && codepoint <= 10FFFF
+    if (codepoint >= 0x10000) {
+        // surrogate pair required
+        inbuf_size = 2;
+        uint32_t code = (codepoint - 0x10000);
+        input_buffer[0] = 0xD800 | (code >> 10);
+        input_buffer[1] = 0xDC00 | (code & 0x3FF);
+    } else {
+        inbuf_size = 1;
+        input_buffer[0] = codepoint;
+    }
+
+    // map freetype encoding to windows codepage
+    switch (encoding) {
+    case FT_ENCODING_MS_SJIS:
+        codepage = 932;
+        break;
+    case FT_ENCODING_MS_GB2312:
+        codepage = 936;
+        break;
+    case FT_ENCODING_MS_BIG5:
+        codepage = 950;
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        codepage = 949;
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        codepage = 1361;
+        break;
+    default:
+        return codepoint;
+    }
+
+    // do convert
+    int output_size = WideCharToMultiByte(
+            codepage,
+            0,
+            input_buffer,
+            inbuf_size,
+            output_buffer,
+            4,
+            NULL,
+            &conversion_fail);
+    if (output_size == 0 || conversion_fail) {
+        return codepoint;
+    }
+
+    result = 0;
+    for (size_t i = 0; i < output_size; ++i) {
+        result <<= 8;
+        result |= (uint8_t)output_buffer[i];
+    }
+
+    return result;
+}
+#elif defined(CONFIG_ICONV)
+static const char* SJIS_ENCODING[] = {
+    "CP932",
+    "SHIFT_JIS",
+    "SHIFTJIS",
+    "SJIS",
+    NULL
+};
+
+static const char* GB2312_ENCODING[] = {
+    "CP936",
+    "GB2312",
+    "GB18030",
+    "GBK",
+    NULL
+};
+
+static const char* BIG5_ENCODING[] = {
+    "CP950",
+    "BIG5",
+    NULL
+};
+
+static const char* WANSUNG_ENCODING[] = {
+    "CP949",
+    NULL
+};
+
+static const char* JOHAB_ENCODING[] = {
+    "CP1361",
+    "JOHAB",
+    NULL
+};
+
+static const char* UCS4_ENCODING[] = {
+    "UCS-4LE",
+    "UCS4LE",
+    "UTF32LE",
+    "UTF-32LE",
+    NULL
+};
+
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    uint32_t result = codepoint;
+    char input_buffer[4];
+    char output_buffer[4];
+    char* inbuf = (char*)&input_buffer;
+    char* outbuf = (char*)&output_buffer;
+    size_t inbuf_size = sizeof(input_buffer);
+    size_t outbuf_size = sizeof(output_buffer);
```

Comment:
Some of the assignments could be moved closer to their first usage.

Reply:
Sorry, I meant to say ”declarations“ not “assignments”. You did already move the assignments; can you also move the declarations?

Reply:
Done.